source/common/http/utility.cc

```diff
@@ -546,6 +567,48 @@ Utility::QueryParams Utility::parseParameters(absl::string_view data, size_t sta
   return params;
 }
 
+Utility::QueryParamsMulti Utility::QueryParamsMulti::parseParameters(absl::string_view data,
+                                                                     size_t start,
+                                                                     bool decode_params) {
+  QueryParamsMulti params;
+
+  while (start < data.size()) {
+    size_t end = data.find('&', start);
+    if (end == std::string::npos) {
+      end = data.size();
+    }
+    absl::string_view param(data.data() + start, end - start);
+
+    const size_t equal = param.find('=');
+    if (equal != std::string::npos) {
+      const auto param_name = StringUtil::subspan(data, start, start + equal);
+      const auto param_value = StringUtil::subspan(data, start + equal + 1, end);
+      params.add(decode_params ? PercentEncoding::decode(param_name) : param_name,
+                 decode_params ? PercentEncoding::decode(param_value) : param_value);
+    } else {
+      const auto param_name = StringUtil::subspan(data, start, end);
+      params.add(decode_params ? PercentEncoding::decode(param_name) : param_name, "");
+    }
+
+    start = end + 1;
+  }
+
+  return params;
+}
+
+void Utility::QueryParamsMulti::remove(absl::string_view key) { this->data.erase(key); }
+
+void Utility::QueryParamsMulti::add(absl::string_view key, const std::string& value) {
+  auto result = this->data.emplace(key, std::vector<std::string>{value});
```

Comment:
For storing in container, it seems safer to extend the lifetime of both key and value.
```
auto result = this->data.emplace(std::string(key), std::vector<std::string>{std::string(value)});
```
string_view to string triggers copy but this copy is needed as we need to store them into container. Your initial approach that passes the arg by value  will trigger copy 2 times (1 is pass by value , the other is here store into map), while the new interface is only one time.

Reply:
Well, now I feel dumb.  Suggested changes made.